zkml/docs/src/llms-layers/qkv.md

```diff
@@ -0,0 +1,95 @@
+
+Given the 3 claims $y_Q$, $y_K,$ and $y_V$, the prover computes also the claims $B_q(r''_Q)$, $B_k(r''_K)$ and $B_v(r''_V)$, and claims $\widetilde{y_Q} = y_Q - B_q(r''_Q)$, $\widetilde{y_K} = y_K - B_k(r''_K)$, $\widetilde{y_V} = y_V - B_V(r''_V)$. The claims $B_q(r''_Q)$, $B_k(r''_K)$ and $B_v(r''_V)$ are then accumulated to be later proven with an opening proof for the corresponding committed bias polynomial.
+
+Given the 3 claims $y'_Q$, $y'_K$, $y'_V$, the verifier samples a random challenge $\lambda$ and then the prover employs a sum-check protocol to prove the relationship:
+\begin{equation}
+\scriptsize
+\widetilde{y_Q} + \lambda \widetilde{y_K} + \lambda^2 \widetilde{y_V} = \sum_{x \in \{0,1\}^{\log(e)}} X(r'_Q, x)*W_Q(x, r''_Q) + \lambda X(r'_K, x)*W_K(x, r''_K) + \lambda^2 X(r'_V, x)*W_V(x,r''_V)
+\tag{1}
+\end{equation}
+
+Where $X$ is the MLE of the input matrix `X`. At the end of the sum-check protocol, for a random $r \in \mathbb{F}^{\log(e)}$ the following claims are produced:
+
+- 3 claims $X(r'_Q, r)$, $X(r'_K, r)$, $X(r'_V, r)$, which can be accumulated in a single claim for the MLE polynomial $X$
+- 3 Claims $y_{W_Q} = W_Q(r, r''_Q)$, $y_{W_K} = W_K(r, r''_K)$, $y_{W_V} = W_V(r, r''_V)$; each of these claims will be added to the set of claims to be opened for the corresponding committed weight polynomial
+
+## Caching Optimization
+In a LLM like GPT2, the inference is an iterative process: the same layers are evaluated multiple times, producing each time an output token. In all iterations except the first one, the input matrix $X_i$ of the QKV layer is the row-wise concatenation of the input matrix $X_{i-1}$ employed in the previous iteration and the vector $x$ representing the output token of the previous iteration. Therefore, also the `Q`, `K` and `V` matrices computed by the QKV layer are a concatenation of the corresponding matrices computed in the previous iteration and the vectors `q`, `k` and `v` computed for the new input row `x`. 
+
+To avoid the cost of a full matrix multiplication in each iteration after the first one, it is thus possible to compute only the vectors `q`, `k` and `v` from the input row `x`, and then concatenate each vector with the corresponding matrices produced by the QKV layer in the previous iteration. 
+
+Note that concatenation is not necessary for the matrix Q computed by the QKV layer: instead, the layer can simply outputs the vector q computed for the new input row x 
```

Comment:
It's more than a note or something not necessary. This is what we wanna do, so we should be very explicit about that.

Reply:
Rephrased in commit https://github.com/Lagrange-Labs/deep-prove/pull/123/commits/a8f86fba6856eed02b5775488b89e187fde4c33b